### ISSUE TITLE
Adams Bashforth startup is broken

### ISSUE THREAD

**Issue by rabernat** [rabernat/pyqg#6]
I think there is an error in the way Adams Bashforth time stepping gets started up.

In the original code, the AB timestep coefficients get initialized in the __init__method with values that make it do forward Euler
```python
         self.dt0 = self.dt
         self.dt1 = 0.
```

Then later in step_forward, _after_ the first AB tendency has been calculated, they get updated.
```python
         # add time tendencies (using Adams-Bashforth):
         self.qh1 = self.filtr*(
                     self.qh1 + self.dt0*self.dqh1dt + self.dt1*self.dqh1dt_p)
         self.qh2 = self.filtr*(
                     self.qh2 + self.dt0*self.dqh2dt + self.dt1*self.dqh2dt_p)  
         
         # remember previous tendencies
         self.dqh1dt_p = self.dqh1dt.copy()
         self.dqh2dt_p = self.dqh2dt.copy()
        
        # the actual Adams-Bashforth stepping can only be used starting
        # at the second time-step and is thus set here:   
        if self.tc==0:
            self.dt0 = 1.5*self.dt
            self.dt1 = -0.5*self.dt
```

Starting with this commit (f7f50f8b520123327f9c4f8ae29b1d7ad74c6a2a), the code to update dt0 and dt1 was moved _before_ the first tendency calculation. This means that the first is calculated using forward Euler, but with 1.5*dt instead of dt. Later, when third order was implemented, the bug persisted, meaning that third order coefficients are used for the second timestep.

I will fix this along with my general rewrite of the AB code.

**Comment by crocha700** [comment]
You're right. I'm sure this is a mistake – definitely my fault. Apologies for breaking this startup. Unfortunately I did not realize because of course this did not change anything anything in the final solutions of the example problem. Perhaps this advocates for the need to have many, unambiguous,  tests...

**Comment by MFJansen** [comment]
Why would this not change the test case? I would think that it should (very little, but probably enough to make the test fail). Or are you referring to the other tests that you added?

**Comment by crocha700** [comment]
I was just referring to the example that is integrating over many years; the change to the equilibrated solutions was insignificant...

The "twolayer" test was failing for me from the beginning probably because of differences in platforms. So, after consulting Ryan, I didn't bother trying to pass it after my changes. Had the test passed when I first forked pyqg, I would have worked to try to make my changes pass that test before submitting the pull request. I think this was the origin of all the mess my changes caused. Sorry guys.

**Comment by rabernat** [comment]
First of all, Cesar, no apologies at all, please. We are sorting this thing out. Lesson learned: be serious about tests! This was my mistake, not yours.

I have fixed this AB issue in the following commit: 1f8c816cb0ab33d296f6471c8f1af5ff12b81781

AB3 is still the default, but there is now an option useAB2 which force only second order to be used. I am now trying to get the test working using the origin values. This is complicated though by the [other issue](https://github.com/rabernat/pyqg/issues/5) which is breaking the new code on the only machine I have which can pass the original tests.

**Comment by MFJansen** [comment]
Nobody is blaming anybody! I'm sorry if it came out that way! We just need to make sure that we have a protocol in place to make sure for the future that we are not putting in bugs and that we agree before making changes that change answers. So we need a good test. I just quickly checked, and it seems like the "twolayer" test is sensitive to this small error in the first timestep. It changes the answer for "qnorm" in the fourth significant digit. That gives an example of something that we do want to  be able to detect with the tests.

**Comment by rabernat** [comment]
Malte, did you check with my latest commit? That should have fixed the AB bug and also restored the second-order AB timestepping to the test case. So it should, in theory, give the same result as the code before Cesar touched it.

**Comment by MFJansen** [comment]
As I just said in the other issue thread: the latest commit is crashing on my laptop:
ValueError: operands could not be broadcast together with shapes (32,17) (32,32)

**Comment by crocha700** [comment]
Guys, I found  thread that was open in "Anaconda-issues" last September. People report this bug with mkl real fft and other bugs!

Here it works in my anaconda installation using mklfft with fftpack:

<pre><code>
In [2]: np.fft.rfft2
Out[2]:  function mklfft.fftpack.rfft2
In [5]: np.fft.rfft2(np.zeros((16,16))).shape
Out[5]: (16, 9)
</code></code>

**Comment by rabernat** [comment]
The timestepping issue was fixed with 1f8c816cb0ab33d296f6471c8f1af5ff12b81781.
